src/EditorFeatures/Core/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -133,7 +134,7 @@ protected internal virtual ImmutableArray<DiagnosticDataLocation> GetLocationsTo
                         //    editorSnapshot.
 
                         var diagnosticSpans = this.GetLocationsToTag(diagnosticData)
-                            .Select(location => GetDiagnosticSnapshotSpan(location, editorSnapshot, sourceText));
+                            .Select(loc => loc.UnmappedFileSpan.GetClampedTextSpan(sourceText).ToSnapshotSpan(snapshot));
```

Comment:
basic idea is that the client decides if they want to use UnmappedFileSpan or MappedFileSpan (ideally with a comment explaining why they are doing that) and then they decide on how to interpret that span.  Normally "clamping" it to some particular source-text instance that can give tehm a valid TextSpan back.